### PR TITLE
HIVE-25953: Drop HiveRelMdPredicates::getPredicates(Join...) to use that of RelMdPredicates

### PR DESCRIPTION
**Pull request by asolimando**
### What changes were proposed in this pull request?

Get rid of the Hive fork of Calcite code.

### Why are the changes needed?

Re-use Calcite code and remove the fork.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing tests. 4 qtests out updated with minor changes.

### CODE REVIEW

File: ql/src/test/results/clientpositive/llap/subquery_scalar.q.out [hunk 1]
```diff
@@ -491,25 +484,25 @@ STAGE PLANS:
...
                 mode: mergepartial
                 outputColumnNames: _col0
-                Statistics: Num rows: 1 Data size: 196 Basic stats: COMPLETE Column stats: NONE
-                Limit
-                  Number of rows: 0
-                  Statistics: Num rows: 0 Data size: 0 Basic stats: COMPLETE Column stats: NONE
+                Statistics: Num rows: 1 Data size: 192 Basic stats: COMPLETE Column stats: NONE
+                Filter Operator
+                  predicate: sq_count_check(_col0) (type: boolean)
+                  Statistics: Num rows: 1 Data size: 192 Basic stats: COMPLETE Column stats: NONE
```
Comment (zabetak) [inline ql/src/test/results/clientpositive/llap/subquery_scalar.q.out:490]: It would be nice to understand why these changes happen.

**Review by zabetak — COMMENTED** [review]
Thanks for the PR @asolimando . Some plan changes are not easy to understand why they took place and if they make the plans better. Do you have some explanations for those?

File: ql/src/test/results/clientpositive/llap/subquery_scalar.q.out (hunk 1 above)
Comment (asolimando) [reply ql/src/test/results/clientpositive/llap/subquery_scalar.q.out:490]: `HiveSortLimit(fetch=[0])` is introduced when `HiveMergeFilterRule` is invoked here:

```
HiveFilter(condition=[false])
  HiveFilter(condition=[sq_count_check($0)])
    HiveAggregate(group=[{}], cnt=[COUNT()])
      HiveProject(p_name=[null:VARCHAR(2147483647) CHARACTER SET "UTF-16LE"])
        HiveFilter(condition=[IS NULL($1)])
          HiveTableScan(table=[[default, part_null_n1]], table:alias=[part_null_n1])
```

The empty filter is rewritten by `HiveRelBuilder` as a LIMIT(0) expression (see [HiveRelBuilder.java#L133](https://github.com/apache/hive/blob/1c3ecf95f94a0329626a9fe46670586e7767f014/ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/HiveRelBuilder.java#L133)).

The "false" predicate is not pushed transitively in this case for the Calcite implementation, so this transformation never happens, we are left with what is the original plan after decorrelation (for that specific subtree):

```
2022-02-22T08:29:38,455 DEBUG [6a4314fb-ef37-46d8-8331-338fff02f1df main] parse.CalcitePlanner: Plan after decorrelation:
HiveProject(p_partkey=[$0], p_name=[$1], p_mfgr=[$2], p_brand=[$3], p_type=[$4], p_size=[$5], p_container=[$6], p_retailprice=[$7], p_comment=
[$8])
  HiveProject(p_partkey=[$0], p_name=[$1], p_mfgr=[$2], p_brand=[$3], p_type=[$4], p_size=[$5], p_container=[$6], p_retailprice=[$7], p_commen
t=[$8], BLOCK__OFFSET__INSIDE__FILE=[$9], INPUT__FILE__NAME=[$10], ROW__ID=[$11], ROW__IS__DELETED=[$12])
    HiveFilter(condition=[=($1, $14)])
      HiveJoin(condition=[true], joinType=[left], algorithm=[none], cost=[not available])
        HiveJoin(condition=[true], joinType=[inner], algorithm=[none], cost=[not available])
          HiveTableScan(table=[[default, part]], table:alias=[part])
          HiveFilter(condition=[sq_count_check($0)])
            HiveAggregate(group=[{}], cnt=[COUNT()])
              HiveProject(p_name=[$1])
                HiveFilter(condition=[IS NULL($1)])
                  HiveTableScan(table=[[default, part_null_n1]], table:alias=[part_null_n1])
        HiveProject(p_name=[$1])
          HiveFilter(condition=[IS NULL($1)])
            HiveTableScan(table=[[default, part_null_n1]], table:alias=[part_null_n1])
```

`false` predicate is not inferred in Calcite, which does not support "non field predicates" (see [HiveRelMdPredicates.java#L439](https://github.com/apache/hive/blob/1c3ecf95f94a0329626a9fe46670586e7767f014/ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/stats/HiveRelMdPredicates.java#L439)), which are predicates without field reference in it.

I will see if I can craft a similar query for Calcite and file an improvement ticket to support such predicates without field reference.

Comment (asolimando) [reply ql/src/test/results/clientpositive/llap/subquery_scalar.q.out:490]: In Calcite, `false` predicates produces a rewrite of the whole subtree as `LogicalValues(tuples=[[]])`, that's why this case is not handled there. The addition of those two lines to Calcite would be trivial, but since this case can't happen, it cannot be reproduced in Calcite and it's hard to justify the change.

I don't know if this can be considered a blocker, it seems a rather extreme case to me.

WDYT?

Comment (zabetak) [reply ql/src/test/results/clientpositive/llap/subquery_scalar.q.out:490]: I am not very familiar with the code but I assume that "non field predicates" does not necessarily mean only `false`. Isn't it possible to craft a predicate other than `false`; what about `RAND() < 0.5` for instance? 

Unfortunately, I think it is a blocker cause if I understand well it causes performance regressions and the latter can be significant ones. In addition, the fact that this change is mostly a refactoring and not something fixing a bug or adding a feature, means that we don't need to rush it.

Comment (asolimando) [reply ql/src/test/results/clientpositive/llap/subquery_scalar.q.out:490]: I agree that we must be conservative with this. 

I managed to create a filter in line with what you suggested, it makes less sense but it's enough to showcase the improvement (explaining that in downstream projects we can have `false` too) and write a unit-test. 

I will open a Calcite ticket and link the current one as "blocked by". 

I will probably use the `Bug` class to keep track of this dependency and/or write a comment in `HiveRelMdPredicates` to describe the situation.
